File: terminal/mcp/agent-output.js
/**
 * Agent-readable MCP text receipts for the Terminal plugin.
 * Structured payloads stay typed for the UI; text is what the model sees.
 */

const ANSI_CSI_RE = /\u001b\[[0-9;?]*[ -/]*[@-~]/g;
const ANSI_OSC_RE = /\u001b\][^\u0007\u001b]*(?:\u0007|\u001b\\)/g;
const ANSI_MISC_RE = /\u001b[@-Z\\-_]/g;

/**
 * @param {string} text
 * @returns {string}
 */
export function stripAnsi(text) {
  return String(text ?? "")
    .replace(ANSI_OSC_RE, "")
    .replace(ANSI_CSI_RE, "")
    .replace(ANSI_MISC_RE, "");
}

/**
 * @param {Record<string, string|number|boolean|null|undefined>} fields
 * @returns {string}
 */
function headerLines(fields) {
  const lines = [];
  for (const [key, value] of Object.entries(fields)) {
    if (value === undefined) continue;
    lines.push(`${key}=${value === null ? "" : String(value)}`);
  }
  return lines.join("\n");
}

/**
 * @param {object} input
 * @param {boolean} input.ok
 * @param {number|null} input.exitCode
 * @param {string} input.shellKind
 * @param {string} input.shell
 * @param {string} input.cwd
 * @param {boolean} input.timedOut
 * @param {boolean} input.truncated
 * @param {string} input.stdout
 * @param {string} input.stderr
 * @param {string|null} [input.signal]
 * @param {number} [input.durationMs]
 */
export function formatExecText(input) {
  const header = headerLines({
    ok: input.ok,
    exit_code: input.exitCode,
    signal: input.signal ?? "",
    shell: input.shellKind,
    shell_path: input.shell,
    cwd: input.cwd,
    timed_out: Boolean(input.timedOut),
    truncated: Boolean(input.truncated),
    ...(input.durationMs !== undefined ? { duration_ms: input.durationMs } : {}),
  });
  return [
    header,
    "",
    "=== stdout ===",
    String(input.stdout ?? "").replace(/\s+$/u, "") ,
    "=== stderr ===",
    String(input.stderr ?? "").replace(/\s+$/u, ""),
    "",
  ].join("\n");
}

/**
 * @param {object} input
 */
export function formatPtyReadText(input) {
  const body = input.ansiStripped
    ? stripAnsi(input.stdout ?? "")
    : String(input.stdout ?? "");
  const header = headerLines({
    ok: true,
    session_id: input.sessionId,
    exited: Boolean(input.exited),
    exit_code: input.exitCode,
    truncated: Boolean(input.truncated),
    ansi_stripped: Boolean(input.ansiStripped),
  });
  return [
    header,
    "",
    "=== output ===",
    body.replace(/\s+$/u, ""),
    "",
  ].join("\n");
}

/**
 * @param {object} input
 */
export function formatShellsText(input) {
  const lines = [
    headerLines({
      ok: true,
      platform: input.platform,
      default: input.defaultKind,
      count: Array.isArray(input.shells) ? input.shells.length : 0,
    }),
    "",
    "kind\tpath\tsource",
  ];
  for (const shell of input.shells ?? []) {
    lines.push(`${shell.kind}\t${shell.path}\t${shell.source}`);
  }
  lines.push("");
  return lines.join("\n");
}

/**
 * @param {Record<string, string|number|boolean|null|undefined>} fields
 */
export function formatOkText(fields) {
  return `${headerLines({ ok: true, ...fields })}\n`;
}

/**
 * @param {object} input
 */
export function formatSessionOpenText(input) {
  return formatOkText({
    session_id: input.sessionId,
    shell: input.shellKind,
    shell_path: input.shell,
    cwd: input.cwd,
    cols: input.cols,
    rows: input.rows,
  });
}

/**
 * @param {string} text
 * @param {Record<string, unknown>} structured
 * @param {{ isError?: boolean }} [opts]
 */
export function mcpToolResult(text, structured, opts = {}) {
  const result = {
    content: [{ type: "text", text: String(text) }],
    structuredContent: structured,
  };
  if (opts.isError) result.isError = true;
  return result;
}

File: terminal/mcp/agent_output.go
package main

import (
	"fmt"
	"strings"
)

// Agent-readable MCP text receipts for the Terminal plugin.
// Port of terminal/mcp/agent-output.js.

func headerLines(fields map[string]any) string {
	order := []string{
		"ok", "exit_code", "signal", "shell", "shell_path", "cwd",
		"timed_out", "truncated", "duration_ms", "session_id", "exited",
		"ansi_stripped", "platform", "default", "count",
	}
	var lines []string
	seen := map[string]bool{}
	for _, k := range order {
		if v, ok := fields[k]; ok {
			if v == nil {
				continue
			}
			lines = append(lines, fmt.Sprintf("%s=%v", k, v))
			seen[k] = true
		}
	}
	for k, v := range fields {
		if v == nil {
			continue
		}
		if !seen[k] {
			lines = append(lines, fmt.Sprintf("%s=%v", k, v))
		}
	}
	return strings.Join(lines, "\n")
}

func formatShellsText(data map[string]any) string {
	shells, _ := data["shells"].([]ResolvedShell)
	var lines []string
	lines = append(lines, headerLines(map[string]any{
		"ok":       true,
		"platform": data["platform"],
		"default":  data["defaultKind"],
		"count":    len(shells),
	}))
	lines = append(lines, "", "kind\tpath\tsource")
	for _, sh := range shells {
		lines = append(lines, fmt.Sprintf("%s\t%s\t%s", sh.Kind, sh.Path, sh.Source))
	}
	lines = append(lines, "")
	return strings.Join(lines, "\n")
}

func formatSessionOpenText(data map[string]any) string {
	return headerLines(map[string]any{
		"ok":         true,
		"session_id": data["sessionId"],
		"shell":      data["shellKind"],
		"shell_path": data["shell"],
		"cwd":        data["cwd"],
		"cols":       data["cols"],
		"rows":       data["rows"],
	}) + "\n"
}

func formatPtyReadText(data map[string]any) string {
	body := fmt.Sprintf("%v", data["stdout"])
	if ansi, _ := data["ansiStripped"].(bool); ansi {
		body = stripANSI(body)
	}
	return strings.Join([]string{
		headerLines(map[string]any{
			"ok":            true,
			"session_id":    data["sessionId"],
			"exited":        data["exited"],
			"exit_code":     data["exitCode"],
			"truncated":     data["truncated"],
			"ansi_stripped": data["ansiStripped"],
		}),
		"",
		"=== output ===",
		strings.TrimRight(body, " \t\r\n"),
		"",
	}, "\n")
}

func formatListSessionsText(items []map[string]any) string {
	var lines []string
	lines = append(lines, headerLines(map[string]any{
		"ok":    true,
		"count": len(items),
	}))
	lines = append(lines, "")
	lines = append(lines, "session_id\tshell\tcwd\texited")
	for _, s := range items {
		lines = append(lines, fmt.Sprintf("%v\t%v\t%v\t%v",
			s["sessionId"], s["shellKind"], s["cwd"], s["exited"]))
	}
	lines = append(lines, "")
	return strings.Join(lines, "\n")
}
